orm/serial_bucket.go

```diff
@@ -0,0 +1,384 @@
+package orm
+
+import (
+	"reflect"
+
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/errors"
+)
+
+// SerialModel is implemented by any entity that can be stored using SerialModelBucket.
+// GetID/SetID are used to store and access the Key.
+// The ID is always set to nil before serializing and storing the Value.
+type SerialModel interface {
+	weave.Persistent
+	Validate() error
+	GetID() []byte
+	SetID([]byte) error
+}
+
+// SerialModelSlicePtr represents a pointer to a slice of SerialModels. Think of it as
+// *[]SerialModel Because of Go type system, using []SerialModel would not work for us.
+// Instead we use a placeholder type and the validation is done during the
+// runtime.
+type SerialModelSlicePtr interface{}
+
+// SerialModelBucket is implemented by buckets that operates on SerialModels rather than
+// Objects.
+type SerialModelBucket interface {
+	// One query the database for a single SerialModel instance. Lookup is done
+	// by the primary index key. Result is loaded into given destination
+	// SerialModel.
+	// This method returns ErrNotFound if the entity does not exist in the
+	// database.
+	// If given SerialModel type cannot be used to contain stored entity, ErrType
+	// is returned.
+	One(db weave.ReadOnlyKVStore, key []byte, dest SerialModel) error
+
+	// PrefixScan will scan for all SerialModels with a primary key (ID)
+	// that begins with the given prefix.
+	// The function returns a (possibly empty) iterator, which can
+	// load each SerialModel as it arrives.
+	// If reverse is true, iterates in descending order (highest value first),
+	// otherwise, it iterates in ascending order.
+	PrefixScan(db weave.ReadOnlyKVStore, prefix []byte, reverse bool) (SerialModelIterator, error)
+
+	// ByIndex returns all objects that secondary index with given name and
+	// given key. Main index is always unique but secondary indexes can
+	// return more than one value for the same key.
+	// All matching entities are appended to given destination slice. If no
+	// result was found, no error is returned and destination slice is not
+	// modified.
+	ByIndex(db weave.ReadOnlyKVStore, indexName string, key []byte, dest SerialModelSlicePtr) error
+
+	// IndexScan does a PrefixScan, but on the named index. This would let us eg. load all counters
+	// in order of their count. Or easily find the lowest or highest count.
+	IndexScan(db weave.ReadOnlyKVStore, indexName string, prefix []byte, reverse bool) (SerialModelIterator, error)
```

Comment:
I think creating an API of chainable methods (i.e. `bucket.Index(name).PrefixScan(...).One(dest)` is problematic for testing and wrapping. Especially packages like `migration` would have problem with this.

I wonder if all those methods cannot be reduced to this single one though. What if we had only `IndexScan` method? To find entities in the database we would need to write something like this:
```go
err := orm.ByID(store, db, ID, dest) // Calls store.IndexScan underneath.
err := orm.ByIndex(store, db, ID, dest) // Calls store.IndexScan underneath.

iter, err := store.IndexScan(db, indexName, prefix, reverse)
```

It might be slightly less friendly to sometimes provide functionality as a function, other time as a method. The positive side is that the interface is getting smaller and any wrapping is easier. User has to also understand less (look at `ByID` implementation - it is a `IndexScan` call that takes the first result!)

Reply:
Hmm... The part I am skeptical about is using `IndexScan` for `ByID` calls result as wrapping of the desired object then unwrapping `SerialModelIterator` to get a single value. Extra steps for every single `ByID` call. This trade-off might reduce performance.

Reply:
Fair point with the chainable APIs not being easy to wrap in goland.

Let's leave as it is, just maybe adjust some names.
I don't see that level of `orm.By...` methods being helpful on the API